platform/machine/azure/cluster.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2018 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package azure
+
+import (
+	"fmt"
+	"math/rand"
+	"os"
+	"path/filepath"
+
+	ctplatform "github.com/coreos/container-linux-config-transpiler/config/platform"
+	"github.com/coreos/pkg/capnslog"
+
+	"github.com/coreos/mantle/platform"
+	"github.com/coreos/mantle/platform/api/azure"
+	"github.com/coreos/mantle/platform/conf"
+)
+
+const (
+	Platform platform.Name = "azure"
+)
+
+var (
+	plog = capnslog.NewPackageLogger("github.com/coreos/mantle", "platform/machine/azure")
+)
+
+type cluster struct {
+	*platform.BaseCluster
+	api            *azure.API
+	sshKey         string
+	ResourceGroup  string
+	StorageAccount string
+}
+
+func NewCluster(opts *azure.Options, rconf *platform.RuntimeConfig) (platform.Cluster, error) {
+	api, err := azure.New(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := api.SetupClients(); err != nil {
+		return nil, fmt.Errorf("setting up clients: %v\n", err)
+	}
+
+	bc, err := platform.NewBaseCluster(opts.Options, rconf, Platform, ctplatform.Azure)
+	if err != nil {
+		return nil, err
+	}
+
+	resourceGroup, err := api.CreateResourceGroup("kola-cluster")
+	if err != nil {
+		return nil, err
+	}
+
+	ac := &cluster{
+		BaseCluster:   bc,
+		api:           api,
+		ResourceGroup: resourceGroup,
+	}
+
+	storageAccount, err := api.CreateStorageAccount(resourceGroup)
+	if err != nil {
+		return nil, err
+	}
+	ac.StorageAccount = storageAccount
+
+	_, err = api.PrepareNetworkResources(resourceGroup)
+	if err != nil {
+		return nil, err
+	}
+
+	if !rconf.NoSSHKeyInMetadata {
+		keys, err := ac.Keys()
+		if err != nil {
+			return nil, err
+		}
+		ac.sshKey = keys[0].String()
```

Comment:
is `keys` always going to have at least 1 element?

Reply:
I believe so, when a new SSH agent (which is what the `Keys` function is listing) is created a key is always generated (see `NewSSHAgent` in `network/ssh.go`). We're also doing this same direct access of `keys[0]` in DigitalOcean.